chips/sam4l/src/pm.rs

```diff
@@ -487,22 +500,69 @@ macro_rules! mask_clock {
         let val = (*PM).$field.get() | ($mask);
         (*PM).$field.set(val);
     });
+
+    ($module:ident: $field:ident & $mask:expr) => ({
+        unlock(concat_idents!($module, _MASK_OFFSET));
+        let val = (*PM).$field.get() & ($mask);
+        (*PM).$field.set(val);
+    });
+}
+
+// Clock masks that allow us to go into deep sleep without disabling any active
+// peripherals.
+
+// FLASHCALW clocks and APBx clocks are allowed
+//
+// This is identical to the reset value of the HSBMASK except it allows the
+// PicoCache RAM clock to be on as well.
+const DEEP_SLEEP_HSBMASK: u32 = 0x1e6;
+
+// No clocks allowed on PBA
+const DEEP_SLEEP_PBAMASK: u32 = 0x0;
+
+// FLASHCALW and HRAMC1 clocks allowed
+//
+// This is identical to the reset value of the PBBMASK except it allows the
+// flash's HRAMC1 clock as well.
+const DEEP_SLEEP_PBBMASK: u32 = 0x3;
+
+/// Determines if the chip can safely go into deep sleep without preventing
+/// currently active peripherals from operating.
+///
+/// We look at the PM's clock mask registers and compare them against a set of
+/// known masks that include no peripherals that can't operate in deep
+/// sleep (or that have no function during sleep).
+///
+/// This means it is the responsibility of each peripheral to disable it's clock
+/// mask whenever it is idle.
+///
+/// We also special case GPIO (which is in PBCMASK), and just see if any interrupts are pending
```

Comment:
Are you saying we can't go into sleep if we want GPIO interrupts? That seems strange to me.